### PR TITLE
feat(gatsby-design-tokens): v2

### PR DESCRIPTION
**Pull request by fk**
## Description

This PR _drafts_ the next major version (2) of `gatsby-design-tokens`.  
TL;DR, we want to

- adjust tokens following suggestion from https://github.com/gatsby-inc/gatsby-interface/issues/181
- add a base `theme-ui` Gatsby theme, and the gatsbyjs.org-specific `theme-ui` theme; also ref. #19115; wondering if those should be separate packages?

## Documentation / Changelog

- [Preview of the updated README](https://github.com/gatsbyjs/gatsby/blob/8eb8b6c12c94453c9467ffa07aaadfcc064137d4/packages/gatsby-design-tokens/README.md)

### Tokens

This update, apart from the goals mentioned above, moves the rudimentary tokens even closer to our current sole target, CSS. This e.g. means that instead of exporting an array of integers for `fontSizes` or `space`, we now provide `rem` values instead (while still making the v1 variants available — but we do have "soft" preferences irt default units for CSS which are emphasized by these breaking changes).  
We currently only consume these tokens in the context of CSS, so let's make things a bit easier there:

- `fontsSizes` exports `rem` values now; old values available at `fontSizesRaw`
- `space` exports `rem` values now, too; old stuff available at `spaceRaw`
- `fonts` exports a string now, ready to be used with the `font-family` CSS prop; old array of font family names available as `fontsLists`
- `radii` now contains the `px` unit

The update also embraces `theme-ui` a bit more directly by partially adopting and/or inheriting its naming conventions for a bunch of tokens, making "theme composition using token groups" even easier than before.

Regarding the discussion in https://github.com/gatsby-inc/gatsby-interface/issues/181, everything but the `space` scale has been adjusted to hopefully "just work", or work easier than before. The update to v2 of `gatsby-design-tokens` should require almost no refactoring of values in `gatsby-interface` —

- the `transition` properties partially follow a different naming convention
- `fonts.header` changed to `fonts.headling`
- `radii` values now come with the `px` unit, so the latter is not needed anymore in template literals

— and apart from that should be solved by removing the transformation of `fontSizes` and `spaces` `to `rem`, and the `join` of `fonts` stacks. (cc @amberleyromo @greglobinski @Elanhant — please also see the "Infamous last words" section at the end).

Here's all changes listed by "token group":

#### `breakpoints`

- BREAKING CHANGE: Remove `breakpoints.xxs`
  - All tokens started as only being consumed via `styled-system`. In its context, and using an object to define breakpoints, `breakpoints.xxs` made sense/was required. Since we switched to `theme-ui`, it doesn't anymore. `gatsby-interface` defines its own set of `breakpoints`; the `www` source already adjusted this when composing its `theme-ui`theme, so this shouldn't hurt at all.
- feat: add export `breakpointsArray`, ready for `theme-ui`'s [responsive styles](https://theme-ui.com/getting-started/#responsive-styles) feature

#### `colors`

- feat: add missing steps in `colors.blackFade` and `colors.whiteFade`:
  - add `colors.blackFade[90]`
  - add `colors.blackFade[40]`
  - add `colors.blackFade[20]`
  - add `colors.blackFade[5]`
  - add `colors.whiteFade[90]`
  - add `colors.whiteFade[40]`
  - add `colors.whiteFade[20]`
  - add `colors.whiteFade[5]`
- BREAKING CHANGE: fix `colors.blackFade[80]`, `colors.whiteFade[80]`
  - both were set to an opacity of `.85`, adjusted to `.8`
  - TODO: Adjust a couple of values for gatsbyjs.org; no clue about where we currently might use this in `gatsby-interface` and related (.com/Cloud dashboard).
- feat: port `colors.code` from the gatsbyjs.org `theme-ui` theme to the `colors` tokens
  - 💚 This improves compliance with WCAG 2.0 AA color contrasts.
- feat: add `colors.code.copyButton`
- feat: add `colors.code.lineHighlightBackground`
- feat: add `colors.code.scrollbarTrack`

#### `fonts`

- feat: Add aliases `body` and `sans` for `fonts.system`
- BREAKING CHANGE: rename `fonts.header` to `fonts.heading`
  - Comply with `theme-ui` requirements: https://theme-ui.com/theme-spec#typography.
  - TODO: Consumers need to refactor `fonts.header` to `fonts.heading`.
- feat: Add alias `brand` for `fonts.heading`
- BREAKING CHANGE: `fonts` exports a string now, ready to be used with the `font-family` CSS prop; the array of font family names is still available as `fontsLists`
- 💚 No need to `join` font family names in `gatsby-interface` any longer.

#### `fontSizes`

- BREAKING CHANGE: `fontSizes` does now contain `rem` values; old scale (integers) still available as `fontSizesRaw`
- feat: add export `fontSizesPx`
- 💚 No need to recalculate to `rem` in `gatsby-interface` anymore.

#### `fontWeights`

- BREAKING CHANGE: Name `fontWeights` — switch from array to object
  - Include the required theme-ui defaults `body`, `heading`, and `bold` (https://theme-ui.com/theme-spec#typography).
  - `body: 400`
  - `semiBold: 600`
    - TODO: gatsbyjs.org's source code expects this to be `medium`, needs minor refactor
  - `bold: 700`
    - TODO: `gatsby-interface` defines `bold` as `800`, whatever it consumes needs to refactor to use `extraBold` instead
  - `heading: 700`
  - `extraBold: 800`
    - TODO: gatsbyjs.org's source code expects this to be `headingPrimary`, needs minor refactor

#### `lineHeights`

- feat: Add line heights `heading` (equivalent to the existing `dense`) and `body` (equivalent to the existing `default`) to satisfy `theme-ui`'s ["Typography" theme requirements](https://theme-ui.com/theme-spec/#typography)

#### `radii`

- BREAKING CHANGE: Add `px` unit to radii scale values.
- 💚 No need to add the `px` unit in when used in template literals, etc.
- TODO: Consumers need to remove `px` from usages in template literals, etc.

#### `sizes`

- BREAKING CHANGE: removed
  - As discussed in https://github.com/gatsby-inc/gatsby-interface/issues/181, this moves the current gatsbyjs.org-centric `sizes` tokens to `theme-gatsbyjs-org` for clarity, until a clear pattern emerges.

#### `space`

- BREAKING CHANGE: `space` does now contain `rem` values; old scale (integers) still available as `spaceRaw`
- feat: add export `spacePx`
- 💚 No need to recalculate to `rem` in `gatsby-interface` anymore.

#### `transition`

- feat: Add `transition.default` shortcut
- feat: Add `transition.curve.fastOutLinearIn` from `gatsby-interface`
- feat: Add `transition.speed.faster` (`50ms`), matching `gatsby-interface`'s `blink`
- feat: Add `transition.speed.slower` (`1000ms`), matching `gatsby-interface`'s `snail`
- BREAKING CHANGE: change `transition.speed.slow` from `350ms` to `500ms`, now matching `gatsby-interface`
- 💚 🔶  All values are now matching `gatsby-interface`, _naming of two properties differs_.
- TODO: Adjust a couple gatsbyjs.org things, would suggest refactoring `gatsby-interface` and related.

#### `zIndices`

- BREAKING CHANGE: removed
  - As discussed in https://github.com/gatsby-inc/gatsby-interface/issues/181, this moves the current gatsbyjs.org-centric `zIndices` tokens to `theme-gatsbyjs-org`for clarity, until a clear pattern emerges.

### Themes

Tested and working with `gatsby-plugin-theme-ui@^0.3.0`, `theme-ui@^0.3.1`; current `www` version of the two is assumed to be working, too. I have a version of `www` working with `gatsby-design-tokens-next` that I'll get a PR up for. — _Edit: see PR #21276_

- feat: add base `theme-ui` Gatsby theme, gatsbyjs.org `theme-ui` theme
  - Both themes don't tree-shake yet, so no `agadoo` yet in the corresponding `gatsby-design-tokens/package.json` scripts. @wardpeet I'd love to take your offer and would appreciate if we could take a look at the mess I made next week! 🙏
  - Both export the `theme-ui` theme as `theme` and also provide individual token group exports (because we require them in `www` for usage outside of the `theme-ui` context).
  - Example usage in `www` — see PR #21276:
    - `export { theme as default } from "gatsby-design-tokens/dist/theme-gatsbyjs-org"
` in `www/src/gatsby-plugin-theme-ui/index`
    - Directly import individual token groups with `import { mediaQueries } from "gatsby-design-tokens/dist/theme-gatsbyjs-org"`.

## Related Issues

- #19115 [www][gatsby-design-tokens] Port tokens / consolidate/add theme-ui theme from `www`
- https://github.com/gatsby-inc/gatsby-interface/issues/181 `gatsby-design-tokens` update
- #21276 chore(www): Adjust for `gatsby-design-tokens` v2

## Infamous last words

- move `gatsby-design-tokens` to `gatsby-interface`?
- add TS definitions right in the tokens package`?
- move themes to bespoke packages?
- `gatsby-interface` `TONES` vs "role-based tokens" in `gatsby-design-tokens`, and `theme-ui`'s [variants](https://theme-ui.com/guides/variants/) — e.g. would be nice to have the `TONES` available for usage with the themes, and `theme-ui`'s `sx` prop, `variants` seems great to spec components (even if we don't use [`theme-ui` components](https://theme-ui.com/components)), etc.
- transformers (px to rem, hex to rgb(a)) are cool :P
- irt `colors`, we already [have a need](https://www.gatsbyjs.org/guidelines/color/) for a bit of [functionality like e.g. this](https://amzn.github.io/style-dictionary/) around additional metadata

### CODE REVIEW

**Comment by fk** [comment]
Ay, this is missing a youtube link: https://www.youtube.com/watch?v=zhVgbZdMdb0

**Comment by sidharthachatterjee** [comment]
Published in `gatsby-design-tokens@2.0.0`

**Comment by fk** [comment]
Thank you @sidharthachatterjee! 💜 

To everyone following along: Let's continue the discussion about further changes over at https://github.com/gatsby-inc/gatsby-interface/issues/181 — the 2.0 release is by no means meant to be a stop to that (, and moar updates). 🤗 

cc @wardpeet The changed `microbundle` scripts seem to have at least worked ("worked well" is another thing 😉) — still would love to take another look together at some point! 🙏

**Comment by fk** [comment]
FWIW — botched updating the changelog — my commit messages (that, squashed, became the PR commit message body) didn't contain the required things, only my PR description had these, and we missed c&p that when merging.

Will get a follow-up PR in. ✌️

**Comment by fk** [comment]
For anyone following along, unfortunately the `2.0.0` package is missing `dist/theme` and `dist/theme-gatsbyjs-org` (see https://unpkg.com/browse/gatsby-design-tokens@2.0.0/dist/ — thanks @sidharthachatterjee)!

I have my guesses on why that is (or … hold on, `dist/index.esm.js` also seems missing — well maybe I'm completely clueless 😛). Will figure out tomorrow — hoping for @sidharthachatterjee to help out the `doge` in me. 🤞 🤗

**Comment by Elanhant** [comment]
@fk 
> move gatsby-design-tokens to gatsby-interface?

If we ever use `gatsby-design-tokens` as part of `gatsby-interface`, then we could do this. Otherwise I'm not sure, since I don't think we want people having to pull the entire gatsby-interface library to get just the design tokens.

> add TS definitions right in the tokens package`?

This makes sense, I can try tackling this when I have a bit more free time :)

**Comment by fk** [comment]
> If we ever use gatsby-design-tokens as part of gatsby-interface, then we could do this. Otherwise I'm not sure, since I don't think we want people having to pull the entire gatsby-interface library to get just the design tokens.

👍 Yeah TBH I can't really make sense of this myself anymore (multiple entry points/tree-shaking aside) — I probably thought `gatsby-interface` was its own GitHub org for a minute there 🤷‍♂

**Comment by fk** [comment]
> >  add TS definitions right in the tokens package`?

> This makes sense, I can try tackling this when I have a bit more free time :)

I already did look into this a _littlle_, maybe I can get a draft PR up …

**Comment by Elanhant** [comment]
> > > add TS definitions right in the tokens package`?
> 
> > This makes sense, I can try tackling this when I have a bit more free time :)
> 
> I already did look into this a _little_, maybe I can get a draft PR up …

I actually have added some definitions to `gatsby-interface` [here](https://github.com/gatsby-inc/gatsby-interface/blob/master/src/typings.d.ts#L6), so probably the easiest approach for now is just to add an `index.d.ts` file with those definitions (updated to reflect latest changes) and make sure it's included in the bundle and correctly linked via `types` property in `package.json` 🙂
